rhubarb/src/audio/OggOpusFileReader.cpp

```diff
@@ -0,0 +1,138 @@
+#include "OggOpusFileReader.h"
+#include "tools/fileTools.h"
+#include "tools/tools.h"
+#include <format.h>
+#include <fstream>
+#include <opusfile.h>
+#include <stdexcept>
+
+using std::ifstream;
+using std::ios_base;
+using std::make_shared;
+using std::filesystem::path;
+
+std::string opusErrorToString(int errorCode) {
+	switch (errorCode) {
+	case OP_EREAD:
+		return "A read from media returned an error.";
+	case OP_EFAULT:
+		return "An internal logic fault; indicates a bug or heap/stack corruption.";
+	case OP_EIMPL:
+		return "Feature not implemented.";
+	case OP_EINVAL:
+		return "Invalid argument.";
+	case OP_ENOTFORMAT:
+		return "Not a recognized format.";
+	case OP_EBADHEADER:
+		return "Invalid header.";
+	case OP_EVERSION:
+		return "Unsupported version.";
+	case OP_ENOTAUDIO:
+		return "Not audio data.";
+	case OP_EBADPACKET:
+		return "Bad packet.";
+	case OP_EBADLINK:
+		return "Bad link.";
+	case OP_ENOSEEK:
+		return "Stream is not seekable.";
+	default:
+		return "An unexpected Opus error occurred.";
+	}
+}
+
+template <typename T>
+T throwOnError(T code) {
+	if (code < 0) {
+		const std::string message = fmt::format("{} (Opus error {})", opusErrorToString(code), code);
+		throw std::runtime_error(message);
+	}
+	return code;
+}
+
+// RAII wrapper around OggOpusFile
+class OggOpusFileHandle final {
+public:
+	OggOpusFileHandle(const path& filePath);
+
+	OggOpusFileHandle(const OggOpusFileHandle&) = delete;
+	OggOpusFileHandle& operator=(const OggOpusFileHandle&) = delete;
+
+	OggOpusFile* get()
+	{
+		return opusFile;
+	}
+
+	~OggOpusFileHandle()
+	{
+		if (opusFile) {
+			op_free(opusFile);
+		}
+	}
+
+private:
+	OggOpusFile* opusFile;
+};
+
+OggOpusFileHandle::OggOpusFileHandle(const path& filePath)
+	: opusFile(nullptr)
+{
+	int error;
+	opusFile = op_open_file(filePath.string().c_str(), &error);
+	throwOnError(error);
+}
+
+OggOpusFileReader::OggOpusFileReader(const path& filePath)
+	: filePath(filePath)
+{
+	OggOpusFileHandle file(filePath);
+
+	const OpusHead* opusHead = op_head(file.get(), -1);
+	// Opus decodes at 48 kHz
+	sampleRate = 48000;
```

Comment:
If the sample rate is always 48kHz, does it make sense to store it in a member variable? Wouldn't a `constexpr` constant work, too?